File: BaekJoon_16928.java
// Baekjoon 16928
// https://www.acmicpc.net/problem/16928

import java.io.*;
import java.util.*;
public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N, M;
        int answer = 0;
        HashMap<Integer, Integer> ladders = new HashMap<>();
        HashMap<Integer, Integer> snakes = new HashMap<>();
        boolean[] visited = new boolean[101];

        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        for (int n=0; n<N; n++) {
            st = new StringTokenizer(br.readLine());
            int ls = Integer.parseInt(st.nextToken());
            int le = Integer.parseInt(st.nextToken());
            ladders.put(ls, le);
        }

        for (int m=0; m<M; m++) {
            st = new StringTokenizer(br.readLine());
            int ss = Integer.parseInt(st.nextToken());
            int se = Integer.parseInt(st.nextToken());
            snakes.put(ss, se);
        }

        Deque<Integer> bfs = new LinkedList<>();
        bfs.offer(1);

        while (true) {
            int repeat_num = bfs.size();
            for (int num=0; num<repeat_num; num++) {
                int state = bfs.poll();
                for (int dice=1; dice<=6; dice++) {
                    int next_state = state + dice;
                    if (ladders.containsKey(next_state))
                        next_state = ladders.get(next_state);
                    else if (snakes.containsKey(next_state))
                        next_state = snakes.get(next_state);

                    if (next_state >= 100) {
                        System.out.println(answer + 1);
                        return;
                    }
                    if (!visited[next_state]) {
                        visited[next_state] = true;
                        bfs.offer(next_state);
                    }
                }
            }
            answer++;
        }
    }
}

File: BaekJoon_16928.py
# Baekjoon 16928
# https://www.acmicpc.net/problem/16928

from collections import deque

N, M = map(int, input().split())

ladders = {}
for _ in range(N):
    ls, le = map(int, input().split())
    ladders[ls] = le

snakes = {}
for _ in range(M):
    ss, se = map(int, input().split())
    ladders[ss] = se

visited = [False] * 101
answer = 0
bfs = deque([1])
done = False
while not done:
    repeat_num = len(bfs)
    for _ in " " * repeat_num:
        state = bfs.popleft()
        next_states = [state + i for i in range(1, 7)]
        for next_state in next_states:
            if next_state in ladders:
                next_state = ladders[next_state]
            elif next_state in snakes:
                next_state = snakes[next_state]

            if next_state >= 100:
                done = True
                break

            if not visited[next_state]:
                visited[next_state] = True
                bfs.append(next_state)

        if done:
            break
    answer += 1
print(answer)
